### ISSUE TITLE
[stable/jenkins] Deprecate non-CasC values

### ISSUE THREAD

**Issue by holmesb** [helm/charts#11276]
Now Jenkins supports [Configuration as Code](https://github.com/helm/charts/pull/9057#issuecomment-454182632), the configuration that is currently done by editing config.xml and other xml files should be marked as deprecated. 

I suggest moving non-JCasC configuration to the bottom of the Readme table and as low as poss in the values. Mark them clearly as "deprecated" and add comments that these options will be removed shortly. Also warn the user in Notes.yaml output if these options are still being used.

Then after a grace period to allow users to wean themselves off the xml-munging methods, these options can be removed by a simplification-refactor. 

Need to confirm these are covered by CasC, but I expect the settings to be deprecated will include:
- Master.UseSecurity
- Master.AuthorizationStrategy
- Master.NumExecutors
- Master.SecurityRealm
- Master.ServicePort (? - might be a prereq for CasC)
- Master.SlaveKubernetesNamespace
- Master.SlaveListenerPort
- Master.CSRF.DefaultCrumbIssuer
- Master.ScriptApproval
- Master.JenkinsAdminEmail
- Master.JenkinsUrl (? - might be a prereq for CasC)
- Master.HostName (? - might be a prereq for CasC)
- Master.CLI
- Master.CredentialsXmlSecret
- Agent.*  - All agent values will be deprecated

**Comment by holmesb** [comment]
I've deprecated as many of the items that are covered by CasC as I could find.  Let me know if there are any more.  The only way to reproduce the same configuration as the existing values.yaml is to enable Configuration as Code by default.   In the Readme and NOTES.txt I've added guidance for people who are still using deprecated values explaining how to use CasC instead.   No breaking changes, but some sections of config.yaml generated errors when I commented-out deprecated values, eg Agent so I had to insert `{{- if .Values.Agent }}` condition to prevent. 
I've also:
- Bumped CasC plugin to latest: from v1.5 to v1.6
- Bumped git plugin from 3.9.1 to 3.9.3 to remove vulnerability warning

Let me know if anything is too opinionated or we we still need to use.  @torstenwalter \ @faizanahmad055, I'm not entirely sure the use cases for Master.AdditionalConfig, but confirm you can achieve what AdditionalConfig does using CasC.   Eg, it has [keycloak](https://github.com/jenkinsci/configuration-as-code-plugin/tree/fb682ae6c1f7bd4ccc2d4743297aa53197c9e0cc/demos/keycloak) support.

Hopefully this will put us on course to [simplify](https://github.com/helm/charts/pull/9057#issuecomment-461722527) this chart in due course.

I've not enabled auto-reload, but open to opinions on whether this feature is stable enough now.  We've had no issues with it.  Configuration as Code is part of the Jenkins project.  Without auto-reload, you have to reboot to make config changes, which isn't feasible except in a test environment.  Without auto-reload there's arguably little point in having Config as Code in this chart at all as it's merely "initial setup as code".  Not what the designers of CasC had in mind.  Your mileage may vary.  Anyway this is a bit outside the scope of this issue so I'll open a new one.

**Comment by torstenwalter** [comment]
I had a look at the PR. It's a rather radical approach to reduce the tweaking options like that. For most of the `ConfigScripts` that should not be a problem as they are very short and copy and paste to customize them should not be to hard.

The only issue I see is with the `kubernetes-slave` as now people would have to copy the whole block even for small modification like using a different image or adjusting resource requests / limits.
Also the possibility to run agents in a separate namespace is gone.

**Comment by holmesb** [comment]
Hi @torstenwalter.  I don't understand why changing a kubernetes-slave value (eg image or resources) requires "copying the whole block".  

Re agents in a separate namespaces, are you referring to [this](https://github.com/holmesb/charts/blob/e0faa81d38915f54a1c39c760563ba7ce7e04503/stable/jenkins/values.yaml#L197):
`jenkinsUrl: {{ template "jenkins.fullname" . }}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}`

Whereas currently config.yaml [has](https://github.com/holmesb/charts/blob/0b89e10942960017a708835e9de6e40b3d9a7525/stable/jenkins/templates/config.yaml#L133):
```
{{- if .Values.Master.SlaveKubernetesNamespace }}
          <jenkinsUrl>http://{{ template "jenkins.fullname" . }}.{{.Release.Namespace}}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}</jenkinsUrl>
...
{{- else }}
          <jenkinsUrl>http://{{ template "jenkins.fullname" . }}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}</jenkinsUrl>
...
{{- end }}
```
Currently in this PR, we are only handling the else case.  Perhaps change values.yaml to:
```
jenkinsUrl: {{ template "jenkins.fullname" . }}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}
# Replace the above line with the following if your master is in a different namespace from your agent:
# jenkinsUrl: http://{{ template "jenkins.fullname" . }}.{{.Release.Namespace}}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}

```
? 
Not as intuitive as simply defining SlaveKubernetesNamespace, but defining agents using CasC is far more flexible that the current config.xml-munging method, eg can have multiple agent templates, etc.  Also less maintenance of this chart is needed since users will be empowered to manage their own config.

**Comment by torstenwalter** [comment]
What I meant was that the configuration looks like this:
```
   ConfigScripts:
      ...
      kubernetes-slave: |
        jenkins:
          clouds:
            - kubernetes:
                name: "jenkins-slave"
                serverUrl: https://kubernetes.default
                skipTlsVerify: true
                jenkinsUrl: {{ template "jenkins.fullname" . }}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}
                connectTimeout: 0
                readTimeout: 0
                containerCapStr: 10
                maxRequestsPerHostStr: 64
                retentionTimeout: 5
                templates:
                  - name: "default"
                    label: "jenkins-slave"
                    idleMinutes: "1"
                    activeDeadlineSeconds: "120"
                    slaveConnectTimeout: "1000"
                    volumes:
                      - secretVolume:
                          secretName: mysecret
                          mountPath: "/var/myapp/mysecret"
                    containers:
                      - name: "jnlp"
                        image: "jenkinsci/jnlp-slave:3.27-1"
                        alwaysPullImage: true
                        workingDir: "/home/jenkins"
                        ttyEnabled: false
                        resourceRequestCpu: "200m"
                        resourceLimitCpu: "1000m"
                        resourceRequestMemory: "256Gi"
                        resourceLimitMemory: "2Gi"
```

Now if I want to change `image: "jenkinsci/jnlp-slave:3.27-1"`  then I need to copy `kubernetes-slave: | ...`  block to be able to modifiy the value. Or is there a way to override just that part?
That's no a big deal and definitely reduces maintenace effort, but on the other hand not everyone is experienced with casc. Which is a pleasure to work with compared to the xml alternative.

Would it do any harm if we use the url with the namespace suffix like below?
`jenkinsUrl: {{ template "jenkins.fullname" . }}.{{.Release.Namespace}}:{{.Values.Master.ServicePort}}{{ default "" .Values.Master.JenkinsUriPrefix }}`

In that case we could default to that and even add `namespace: "{{ default .Release.Namespace .Values.Master.SlaveKubernetesNamespace }}"`

**Comment by holmesb** [comment]
> to change `image: "jenkinsci/jnlp-slave:3.27-1"` then I need to copy `kubernetes-slave: | ...` block to be able to modifiy the value. Or is there a way to override just that part?

You just change the image value.  No need to copy the entire block.  I've just validated this in my test Jenkins.

> namespace suffix?

I assumed if there's no namespace this would leave a trailing dot, but in fact, it suffixes `.default`, which is fine.  I've updated.  Nice suggestion.

> 'namespace: "{{ default .Release.Namespace .Values.Master.SlaveKubernetesNamespace }}"'

Not sure how this would work in the case of multiple agent templates in different namespaces.   Letting people manage their own config rather than trying to over-engineer this chart is a good general principle IMO.  Yes "not everyone is experienced with casc", but I think it's intuitive enough and well-documented.  I expect in 99% of cases, people will simply copy and paste from the demos, tweak a few values and be up-and-running.
